File: src/main/java/com/perforce/performance/agent/PageLoadAnalyzerAgent.java
package com.perforce.performance.agent;

import com.perforce.performance.config.ConfigReader;
import com.perforce.performance.driver.DriverManager;
import com.perforce.performance.metrics.MetricsCollector;
import com.perforce.performance.metrics.PerformanceMetrics;
import com.perforce.performance.pages.LoginPage;
import com.perforce.performance.pages.NavigationRouter;
import com.perforce.performance.reports.ExcelReportManager;
import com.perforce.performance.reports.HtmlReportManager;
import org.openqa.selenium.WebDriver;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Date;
import java.util.List;
import java.util.stream.Collectors;

/**
 * PageLoadAnalyzerAgent — Analyzes one or many URLs and produces per-URL HTML reports
 * plus a single combined HTML report (no screenshots) and a combined Excel report.
 *
 * Single URL:
 *   PerformanceAnalysisResult r = PageLoadAnalyzerAgent.analyze("https://example.com");
 *
 * Multiple URLs:
 *   List<PerformanceAnalysisResult> results =
 *       PageLoadAnalyzerAgent.analyzeAll(List.of("https://a.com", "https://b.com"));
 *
 * Metrics captured: Load Time · TTFB · FCP · LCP · TTI · CLS · INP · TBT · DNS · TCP · Resources
 */
public class PageLoadAnalyzerAgent {

    private static final Logger log = LoggerFactory.getLogger(PageLoadAnalyzerAgent.class);

    private PageLoadAnalyzerAgent() {}

    // ─── Single URL ───────────────────────────────────────────────────────────

    /** Analyze one URL — page name derived from hostname + path. */
    public static PerformanceAnalysisResult analyze(String url) {
        return analyze(url, derivePageName(url));
    }

    /** Analyze one URL with an explicit page label. */
    public static PerformanceAnalysisResult analyze(String url, String pageName) {
        log.info("=== PageLoadAnalyzerAgent: {} [{}] ===", pageName, url);

        HtmlReportManager.initForPage(pageName);

        PerformanceMetrics metrics = runMeasurement(url, pageName);

        HtmlReportManager.addMetrics("Page Load Analysis: " + pageName, metrics);
        HtmlReportManager.addSummaryTable(Collections.singletonList(metrics));
        HtmlReportManager.flush();
        String htmlPath = HtmlReportManager.getReportPath();

        String timestamp  = new SimpleDateFormat("yyyy-MM-dd_HH-mm-ss").format(new Date());
        String excelPath  = "reports" + File.separator + "page-load-" + safeName(pageName) + "-" + timestamp + ".xlsx";
        ExcelReportManager.generate(Collections.singletonList(metrics), excelPath);

        logConsoleSummary(metrics);
        log.info("=== Analysis complete. Overall: {} ===", metrics.getOverallStatus());

        return new PerformanceAnalysisResult(metrics, htmlPath, excelPath);
    }

    // ─── Multiple URLs ────────────────────────────────────────────────────────

    /**
     * Analyze a list of URLs.
     * <ul>
     *   <li>Each URL gets its own HTML report (with screenshot).</li>
     *   <li>A combined HTML report (no screenshots) is created for all URLs.</li>
     *   <li>A single combined Excel report covers all URLs.</li>
     * </ul>
     *
     * @param urls List of URLs to analyze (blank entries are skipped).
     * @return     One {@link PerformanceAnalysisResult} per URL, each carrying the individual
     *             HTML path, and the shared combined Excel path.
     */
    public static List<PerformanceAnalysisResult> analyzeAll(List<String> urls) {
        List<String> valid = urls.stream()
                .map(String::trim)
                .filter(u -> !u.isBlank())
                .collect(Collectors.toList());

        log.info("╔══════════════════════════════════════════════════════════╗");
        log.info("║   PageLoadAnalyzerAgent — {} URL(s) to analyze           ", valid.size());
        log.info("╚══════════════════════════════════════════════════════════╝");

        String batchTimestamp = new SimpleDateFormat("yyyy-MM-dd_HH-mm-ss").format(new Date());
        List<PerformanceAnalysisResult> results   = new ArrayList<>();
        List<PerformanceMetrics>        allMetrics = new ArrayList<>();

        int total = valid.size();
        for (int i = 0; i < total; i++) {
            String url      = valid.get(i);
            String pageName = derivePageName(url);
            log.info("[{}/{}] Analyzing: {} — {}", i + 1, total, pageName, url);

            // ── Individual HTML report ────────────────────────────────────────
            HtmlReportManager.initForPage(pageName);

            PerformanceMetrics metrics = runMeasurement(url, pageName);

            HtmlReportManager.addMetrics("Page Load Analysis: " + pageName, metrics);
            HtmlReportManager.addSummaryTable(Collections.singletonList(metrics));
            HtmlReportManager.flush();
            String individualHtmlPath = HtmlReportManager.getReportPath();

            allMetrics.add(metrics);
            logConsoleSummary(metrics);

            results.add(new PerformanceAnalysisResult(metrics, individualHtmlPath, null /* filled below */));
        }

        // ── Combined HTML report (no screenshots) ─────────────────────────────
        HtmlReportManager.initCombined();
        for (PerformanceMetrics m : allMetrics) {
            HtmlReportManager.addMetricsNoScreenshot("Page Load Analysis: " + m.getPageName(), m);
        }
        HtmlReportManager.addSummaryTable(allMetrics);
        HtmlReportManager.flush();
        String combinedHtmlPath = HtmlReportManager.getReportPath();

        // ── Combined Excel report ─────────────────────────────────────────────
        String combinedExcelPath = "reports" + File.separator
                + "combined-analysis-" + batchTimestamp + ".xlsx";
        ExcelReportManager.generate(allMetrics, combinedExcelPath);

        log.info("══════════════════════════════════════════════════════════");
        log.info("  All {} URL(s) analyzed.", allMetrics.size());
        log.info("  Combined HTML  : {}", combinedHtmlPath);
        log.info("  Combined Excel : {}", combinedExcelPath);
        log.info("══════════════════════════════════════════════════════════");

        // Attach combined paths to every result
        return results.stream()
                .map(r -> new PerformanceAnalysisResult(r.getMetrics(), r.getHtmlReportPath(), combinedExcelPath))
                .collect(Collectors.toList());
    }

    // ─── Core measurement ─────────────────────────────────────────────────────

    /**
     * Navigates to {@code url}, handles login if the app redirects to the sign-in
     * page, then waits for the target page to be fully ready before collecting metrics.
     *
     * Flow for authenticated pages (e.g. {@code #/home}, {@code #/ip/catalog}):
     *   1. driver.get(url)  →  app redirects to login
     *   2. Detect login redirect (hash path contains "signin" / "login" / "security")
     *   3. Enter credentials from config.properties and click submit
     *   4. router.waitForCurrentPage() — waits for home page fully ready (header interactive)
     *   5. Inject WebVitals observer (active from the next navigation's first paint)
     *   6. router.navigateAndPrepare(url) — AppNavigator clicks header button(s), then
     *      readiness handler waits for page heading + spinners gone
     *   7. Collect and return metrics
     *
     * Flow for public pages (e.g. the login page itself):
     *   1. driver.get(url)
     *   2. Inject WebVitals observer
     *   3. Reload so observer is active from first paint
     *   4. NavigationRouter.waitForCurrentPage()
     *   5. Collect and return metrics
     */
    private static PerformanceMetrics runMeasurement(String url, String pageName) {
        ConfigReader config = ConfigReader.getInstance();
        WebDriver driver = DriverManager.initDriver();
        try {
            NavigationRouter router    = new NavigationRouter(driver);
            MetricsCollector collector = new MetricsCollector(driver);

            // ── Step 1: navigate to target URL ────────────────────────────────
            driver.get(url);
            String landedUrl  = waitUrlStable(driver);
            String landedPath = NavigationRouter.extractHashPath(landedUrl);

            boolean redirectedToLogin = landedPath.contains("signin")
                    || landedPath.contains("login")
                    || landedPath.startsWith("security");

            if (redirectedToLogin) {
                // ── Step 2: perform login ──────────────────────────────────────
                log.info("[agent] Redirected to login — logging in as '{}'", config.getUsername());
                LoginPage loginPage = new LoginPage(driver);
                loginPage.login(config.getUsername(), config.getPassword());

                // ── Step 3: wait for post-login home page to be fully ready ───
                // This ensures the app header (Catalogs / Administration menus)
                // is interactive before AppNavigator tries to click any button.
                // The home handler waits for: title "Home" + IPs button visible + spinners gone.
                log.info("[agent] Waiting for home page to be fully ready...");
                router.waitForCurrentPage();
                log.info("[agent] Home page ready — URL: {}", driver.getCurrentUrl());

                // ── Step 4: inject observer BEFORE navigating to target ────────
                // Observer must be active so LCP/FCP/CLS events on the target
                // page are captured from the very first render frame.
                collector.injectWebVitalsObserver();

                // ── Step 5: click-based navigation to target ───────────────────
                // AppNavigator performs the same button clicks a real user would:
                //   #/home        → product logo click (already there, no-op via URL)
                //   #/ip/catalog  → Catalogs menu → "IP catalog"
                //   #/library/catalog → Catalogs menu → "Library catalog"
                //   #/query/search    → Advanced search link
                //   #/library/manage  → Administration menu → "Library management"
                //   … etc.
                // NavigationRouter then waits for the page heading + spinner gone.
                log.info("[agent] Navigating to target page: {}", url);
                router.navigateAndPrepare(url);

            } else {
                // ── Public page: inject observer, reload, wait for ready ───────
                collector.injectWebVitalsObserver();
                driver.navigate().refresh();
                router.waitForCurrentPage();
            }

            PerformanceMetrics metrics = collector.collect(pageName);
            metrics.setAgGridLoadTime(router.getLastAgGridLoadTime());
            return metrics;

        } finally {
            DriverManager.quitDriver();
        }
    }

    /** Polls until the URL stops changing (up to ~10 s) and returns the stable URL. */
    private static String waitUrlStable(WebDriver driver) {
        String prev = driver.getCurrentUrl();
        for (int i = 0; i < 15; i++) {
            try { Thread.sleep(200); } catch (InterruptedException e) { Thread.currentThread().interrupt(); break; }
            String cur = driver.getCurrentUrl();
            if (cur.equals(prev)) return cur;
            prev = cur;
        }
        return driver.getCurrentUrl();
    }

    // ─── Console summary ──────────────────────────────────────────────────────

    private static void logConsoleSummary(PerformanceMetrics m) {
        log.info("┌──────────────────────────────────────────────────────────┐");
        log.info("│             PAGE LOAD PERFORMANCE ANALYSIS               │");
        log.info("├──────────────────────────────────────────────────────────┤");
        log.info("│  URL       : {}", m.getPageUrl());
        log.info("│  Page Name : {}", m.getPageName());
        log.info("├──────────────────────────┬────────────┬──────────────────┤");
        log.info("│  Metric                  │   Value    │  Status          │");
        log.info("├──────────────────────────┼────────────┼──────────────────┤");
        log.info("│  Page Load Time          │ {} ms │  {}", rpad(fmt(m.getPageLoadTime()), 7),       pad(m.getPageLoadStatus()));
        log.info("│  TTFB                    │ {} ms │  {}", rpad(fmt(m.getTimeToFirstByte()), 7),    pad(m.getTtfbStatus()));
        log.info("│  FCP                     │ {} ms │  {}", rpad(fmt(m.getFirstContentfulPaint()), 7), pad(m.getFcpStatus()));
        log.info("│  LCP                     │ {} ms │  {}", rpad(fmt(m.getLargestContentfulPaint()), 7), pad(m.getLcpStatus()));
        log.info("│  TTI                     │ {} ms │  {}", rpad(fmt(m.getTimeToInteractive()), 7),  pad(m.getTtiStatus()));
        log.info("│  CLS                     │ {}    │  {}", rpad(String.format("%.4f", m.getCumulativeLayoutShift()), 7), pad(m.getClsStatus()));
        log.info("├──────────────────────────┼────────────┼──────────────────┤");
        log.info("│  DNS Lookup              │ {} ms │                  │", rpad(fmt(m.getDnsLookupTime()), 7));
        log.info("│  TCP Connect             │ {} ms │                  │", rpad(fmt(m.getTcpConnectionTime()), 7));
        log.info("│  DOM Interactive         │ {} ms │                  │", rpad(fmt(m.getDomInteractiveTime()), 7));
        log.info("│  DOM Content Loaded      │ {} ms │                  │", rpad(fmt(m.getDomContentLoadedTime()), 7));
        log.info("│  Total Resources         │ {}    │                  │", rpad(String.valueOf(m.getTotalResources()), 7));
        log.info("│  Transfer Size           │ {} KB │                  │", rpad(String.valueOf(m.getTransferSize() / 1024), 7));
        log.info("├──────────────────────────┴────────────┴──────────────────┤");
        log.info("│  OVERALL STATUS : {}", m.getOverallStatus());
        log.info("└──────────────────────────────────────────────────────────┘");
    }

    // ─── Helpers ──────────────────────────────────────────────────────────────

    private static String fmt(double ms) {
        return ms > 0 ? String.format("%.0f", ms) : "N/A";
    }

    private static String rpad(String s, int width) {
        return String.format("%-" + width + "s", s);
    }

    private static String pad(String status) {
        return status == null ? "N/A             │" : String.format("%-16s│", status);
    }

    static String derivePageName(String url) {
        try {
            java.net.URI uri = new java.net.URI(url);
            String host = uri.getHost();

            // Prefer the hash path for SPA-style URLs (e.g. #/ip/catalog → "ip/catalog")
            String hashPath = NavigationRouter.extractHashPath(url);
            if (!hashPath.isEmpty()) {
                String last = NavigationRouter.extractSegment(url);
                if (!last.isBlank()) {
                    // Capitalise first letter for readability (e.g. "home" → "Home")
                    String label = Character.toUpperCase(last.charAt(0)) + last.substring(1);
                    return host + " / " + label;
                }
            }

            // Fall back to last regular path segment
            String path = uri.getPath();
            if (path != null && !path.isEmpty() && !path.equals("/")) {
                String last = path.replaceAll("/$", "").replaceAll(".*/", "");
                if (!last.isBlank()) return host + "/" + last;
            }

            return host != null ? host : url;
        } catch (Exception e) {
            return url;
        }
    }

    private static String safeName(String name) {
        return name.replaceAll("[^a-zA-Z0-9._-]", "_");
    }

    // ─── Result DTO ───────────────────────────────────────────────────────────

    /**
     * Holds metrics and report file paths for one analyzed URL.
     */
    public static class PerformanceAnalysisResult {

        private final PerformanceMetrics metrics;
        private final String htmlReportPath;
        private final String excelReportPath;

        public PerformanceAnalysisResult(PerformanceMetrics metrics,
                                         String htmlReportPath,
                                         String excelReportPath) {
            this.metrics        = metrics;
            this.htmlReportPath = htmlReportPath;
            this.excelReportPath = excelReportPath;
        }

        public PerformanceMetrics getMetrics()        { return metrics; }
        public String getHtmlReportPath()             { return htmlReportPath; }
        public String getExcelReportPath()            { return excelReportPath; }

        @Override
        public String toString() {
            return String.format("PerformanceAnalysisResult{page=%s, overall=%s, html=%s}",
                    metrics.getPageName(), metrics.getOverallStatus(), htmlReportPath);
        }
    }
}

File: src/agent/PageLoadAnalyzerAgent.ts
import * as path from 'path';
import { BrowserManager } from '../driver/BrowserManager';
import { NavigationRouter } from '../pages/NavigationRouter';
import { MetricsCollector } from '../metrics/MetricsCollector';
import { LighthouseCollector } from '../metrics/LighthouseCollector';
import { NavigationAgent } from './NavigationAgent';
import { LoginPage } from '../pages/LoginPage';
import { PerformanceMetrics } from '../metrics/PerformanceMetrics';
import * as HtmlReportManager from '../reports/HtmlReportManager';
import { ExcelReportManager } from '../reports/ExcelReportManager';
import { ConfigReader } from '../config/ConfigReader';
import * as LogManager from '../utils/LogManager';

export interface PerformanceAnalysisResult {
  metrics: PerformanceMetrics;
  htmlReportPath: string;
  excelReportPath: string;
}

/**
 * PageLoadAnalyzerAgent — standalone agent that analyzes URLs without depending
 * on the Playwright Test fixture lifecycle.  Manages its own browser session.
 *
 * Usage:
 *   const result = await PageLoadAnalyzerAgent.analyze('https://example.com');
 *   const results = await PageLoadAnalyzerAgent.analyzeAll(['https://a.com', 'https://b.com']);
 */
/** Set to true before calling analyze/analyzeAll to skip the Lighthouse step. */
export let skipLighthouseAudit = false;
export function setSkipLighthouse(skip: boolean): void { skipLighthouseAudit = skip; }

/** Optional credential override — if set, used instead of config.properties values. */
let customCredentials: { username: string; password: string } | null = null;
export function setCustomCredentials(creds: { username: string; password: string } | null): void {
  customCredentials = creds;
}

export class PageLoadAnalyzerAgent {
  private PageLoadAnalyzerAgent() {}

  // ── Single URL ─────────────────────────────────────────────────────────────

  static async analyze(url: string, pageName?: string): Promise<PerformanceAnalysisResult> {
    const name      = pageName ?? derivePageName(url);
    const reportDir = ConfigReader.getInstance().getReportPath();
    LogManager.init(reportDir);

    console.info(`=== PageLoadAnalyzerAgent: ${name} [${url}] ===`);

    HtmlReportManager.initForPage(name);

    const metrics = await runMeasurement(url, name);

    HtmlReportManager.addMetrics(`Page Load Analysis: ${name}`, metrics);
    HtmlReportManager.addSummaryTable([metrics]);
    HtmlReportManager.flush();
    const htmlPath = HtmlReportManager.getReportPath();

    const ts         = tsStamp();
    const excelPath  = path.join(reportDir, `page-load-${safeName(name)}-${ts}.xlsx`);
    await ExcelReportManager.generate([metrics], excelPath);

    logConsoleSummary(metrics);
    console.info(`=== Analysis complete. Overall: ${metrics.overallStatus} ===`);
    const logPath = LogManager.flush();
    console.info(`=== Log saved: ${logPath} ===`);
    return { metrics, htmlReportPath: htmlPath, excelReportPath: excelPath };
  }

  // ── Multiple URLs ──────────────────────────────────────────────────────────

  static async analyzeAll(urls: string[]): Promise<PerformanceAnalysisResult[]> {
    const valid     = urls.map(u => u.trim()).filter(u => u.length > 0);
    const reportDir = ConfigReader.getInstance().getReportPath();
    LogManager.init(reportDir);

    console.info('╔══════════════════════════════════════════════════════════╗');
    console.info(`║   PageLoadAnalyzerAgent — ${valid.length} URL(s) to analyze`);
    console.info('╚══════════════════════════════════════════════════════════╝');

    const batchTs    = tsStamp();
    const results:    PerformanceAnalysisResult[] = [];
    const allMetrics: PerformanceMetrics[]         = [];

    for (let i = 0; i < valid.length; i++) {
      const url      = valid[i];
      const pageName = derivePageName(url);
      console.info(`[${i + 1}/${valid.length}] Analyzing: ${pageName} — ${url}`);

      HtmlReportManager.initForPage(pageName);
      const metrics = await runMeasurement(url, pageName);
      HtmlReportManager.addMetrics(`Page Load Analysis: ${pageName}`, metrics);
      HtmlReportManager.addSummaryTable([metrics]);
      HtmlReportManager.flush();
      const individualHtml = HtmlReportManager.getReportPath();

      allMetrics.push(metrics);
      logConsoleSummary(metrics);
      results.push({ metrics, htmlReportPath: individualHtml, excelReportPath: '' });
    }

    // Combined HTML
    HtmlReportManager.initCombined();
    allMetrics.forEach(m => {
      HtmlReportManager.addMetricsNoScreenshot(`Page Load Analysis: ${m.pageName}`, m);
    });
    HtmlReportManager.addSummaryTable(allMetrics);
    HtmlReportManager.flush();
    const combinedHtml = HtmlReportManager.getReportPath();

    // Combined Excel
    const combinedExcel = path.join(reportDir, `combined-analysis-${batchTs}.xlsx`);
    await ExcelReportManager.generate(allMetrics, combinedExcel);

    const logPath = LogManager.flush();

    console.info('══════════════════════════════════════════════════════════');
    console.info(`  All ${allMetrics.length} URL(s) analyzed.`);
    console.info(`  Combined HTML  : ${combinedHtml}`);
    console.info(`  Combined Excel : ${combinedExcel}`);
    console.info(`  Run Log        : ${logPath}`);
    console.info('══════════════════════════════════════════════════════════');

    return results.map(r => ({ ...r, excelReportPath: combinedExcel }));
  }
}

// ── Core measurement ─────────────────────────────────────────────────────────────

async function runMeasurement(url: string, pageName: string): Promise<PerformanceMetrics> {
  const config  = ConfigReader.getInstance();
  const session = await BrowserManager.launch();
  const { page } = session;

  try {
    const router    = new NavigationRouter(page);
    const collector = new MetricsCollector(page);

    const targetHashPath = NavigationRouter.extractHashPath(url);
    const isSigninPage   =
      targetHashPath.includes('signin') ||
      targetHashPath.includes('login')  ||
      targetHashPath.startsWith('security');

    if (isSigninPage) {
      // Signin page — measure it directly (no auth loop needed)
      console.info('[agent] Signin page — measuring directly');
      await collector.injectWebVitalsObserver();
      await page.goto(url, { waitUntil: 'domcontentloaded', timeout: config.getPageLoadTimeout() * 1000 });
      await router.waitForCurrentPage();
    } else {
      // ── Step 1: Always start from the signin page ──────────────────────────
      const loginUrl = config.getLoginUrl();
      console.info(`[agent] → Signin: ${loginUrl}`);
      await page.goto(loginUrl, { waitUntil: 'domcontentloaded', timeout: config.getPageLoadTimeout() * 1000 });
      await router.waitForCurrentPage();

      // ── Step 2: Login ──────────────────────────────────────────────────────
      console.info(`[agent] Logging in as '${customCredentials?.username || config.getUsername()}'`);
      const loginPage = new LoginPage(page, config.getPageLoadTimeout() * 1000);
      const username = customCredentials?.username || config.getUsername();
      const password = customCredentials?.password || config.getPassword();
      await loginPage.login(username, password);

      // ── Step 3: Wait for login redirect to complete ────────────────────────
      // waitForCurrentPage() uses URL-stability polling which returns after the
      // first stable interval — too fast if the login AJAX is still in flight.
      // Explicitly wait until the URL leaves the signin/login/security path.
      console.info('[agent] Waiting for post-login redirect...');
      const timeoutMs = config.getPageLoadTimeout() * 1000;
      try {
        await page.waitForURL(
          u => !u.href.includes('signin') && !u.href.includes('login') && !u.href.includes('security/'),
          { timeout: timeoutMs },
        );
      } catch {
        const stuck = page.url();
        console.warn(`[agent] ⚠ Login redirect timed out — still on "${stuck}"`);
        console.warn(`[agent]   Check credentials in config.properties or enter them in the Navigation Plan.`);
      }
      console.info(`[agent] Post-login URL: ${page.url()}`);

      // ── Step 4: Wait for home page readiness ──────────────────────────────
      console.info('[agent] Waiting for home page...');
      await router.waitForCurrentPage();
      console.info(`[agent] Home ready — ${page.url()}`);

      // ── Step 5: Inject observer BEFORE navigating to target ────────────────
      // Observer must be active from the very first paint of the target page.
      await collector.injectWebVitalsObserver();

      // ── Step 6: Navigate to target via UI actions only (strict — no goto) ──
      console.info(`[agent] Navigating to target via UI actions: ${url}`);
      console.info(`[agent] ${NavigationAgent.describeNavigation(url)}`);
      const navResult = await NavigationAgent.navigateTo(
        page, url, timeoutMs, /* strict= */ true,
      );
      navResult.log.forEach(line => console.info(line));

      // ── Step 7: Wait for page readiness ────────────────────────────────────
      await router.waitForCurrentPage();

      // ── Step 7b: Verify we landed on the target, not back on signin ────────
      const landedUrl = page.url();
      if (landedUrl.includes('signin') || landedUrl.includes('login') || landedUrl.includes('security/')) {
        console.warn(`[agent] ⚠ Auth redirect after navigation: landed on "${landedUrl}"`);
        console.warn(`[agent]   Expected target: "${url}"`);
        console.warn(`[agent]   Cause: invalid credentials, session timeout, or missing permissions.`);
      }
    }

    const metrics = await collector.collect(pageName);
    metrics.agGridLoadTime = router.getLastAgGridLoadTime();
    metrics.actionToLoadMs = router.getLastActionToLoadTime();

    // Extract session cookies for authenticated Lighthouse audits
    const sessionCookies = await session.context.cookies();

    // Run Lighthouse audit (launches its own Chrome — no port conflict with Playwright)
    if (!skipLighthouseAudit) try {
      const outputDir = path.join(config.getReportPath(), 'lighthouse');
      metrics.lighthouseData = await LighthouseCollector.run(url, {
        cookies:   sessionCookies,
        outputDir,
        pageName,
      });
    } catch (lhErr) {
      console.warn(`[agent] Lighthouse audit skipped: ${lhErr}`);
    }

    return metrics;
  } finally {
    await BrowserManager.close(session);
  }
}

// ── URL helpers ────────────────────────────────────────────────────────────────

export function derivePageName(url: string): string {
  try {
    const u = new URL(url);
    const host = u.hostname;
    const hashPath = NavigationRouter.extractHashPath(url);
    if (hashPath) {
      const last = NavigationRouter.extractSegment(url);
      if (last) return `${host} / ${last.charAt(0).toUpperCase()}${last.slice(1)}`;
    }
    const p = u.pathname.replace(/\/$/, '').replace(/.*\//, '');
    if (p) return `${host}/${p}`;
    return host || url;
  } catch {
    return url;
  }
}

// ── Console summary ────────────────────────────────────────────────────────────

function logConsoleSummary(m: PerformanceMetrics): void {
  console.info('┌──────────────────────────────────────────────────────────┐');
  console.info('│             PAGE LOAD PERFORMANCE ANALYSIS               │');
  console.info('├──────────────────────────────────────────────────────────┤');
  console.info(`│  URL       : ${m.pageUrl}`);
  console.info(`│  Page Name : ${m.pageName}`);
  console.info('├──────────────────────────┬────────────┬──────────────────┤');
  console.info('│  Metric                  │   Value    │  Status          │');
  console.info('├──────────────────────────┼────────────┼──────────────────┤');
  console.info(`│  Page Load Time          │ ${rpad(fmt(m.pageLoadTime), 7)} ms │  ${m.pageLoadStatus}`);
  console.info(`│  TTFB                    │ ${rpad(fmt(m.timeToFirstByte), 7)} ms │  ${m.ttfbStatus}`);
  console.info(`│  FCP                     │ ${rpad(fmt(m.firstContentfulPaint), 7)} ms │  ${m.fcpStatus}`);
  console.info(`│  LCP                     │ ${rpad(fmt(m.largestContentfulPaint), 7)} ms │  ${m.lcpStatus}`);
  console.info(`│  TTI                     │ ${rpad(fmt(m.timeToInteractive), 7)} ms │  ${m.ttiStatus}`);
  console.info(`│  CLS                     │ ${rpad(m.cumulativeLayoutShift.toFixed(4), 7)}    │  ${m.clsStatus}`);
  console.info('├──────────────────────────┼────────────┼──────────────────┤');
  console.info(`│  DNS Lookup              │ ${rpad(fmt(m.dnsLookupTime), 7)} ms │`);
  console.info(`│  TCP Connect             │ ${rpad(fmt(m.tcpConnectionTime), 7)} ms │`);
  console.info(`│  DOM Interactive         │ ${rpad(fmt(m.domInteractiveTime), 7)} ms │`);
  console.info(`│  DOM Content Loaded      │ ${rpad(fmt(m.domContentLoadedTime), 7)} ms │`);
  console.info(`│  Total Resources         │ ${rpad(String(m.totalResources), 7)}    │`);
  console.info(`│  Transfer Size           │ ${rpad(String(Math.round(m.transferSize / 1024)), 7)} KB │`);
  console.info('├──────────────────────────┴────────────┴──────────────────┤');
  console.info(`│  OVERALL STATUS : ${m.overallStatus}`);

  // ── Lighthouse section (only when audit ran) ─────────────────────────────
  const lh = m.lighthouseData;
  if (lh) {
    console.info('├──────────────────────────────────────────────────────────┤');
    console.info('│                  LIGHTHOUSE AUDIT                        │');
    console.info('├──────────────────────────┬────────────┬──────────────────┤');
    console.info(`│  Performance Score       │ ${rpad(String(lh.performanceScore), 7)}    │  ${scoreLabel(lh.performanceScore)}`);
    console.info(`│  Accessibility Score     │ ${rpad(String(lh.accessibilityScore), 7)}    │  ${scoreLabel(lh.accessibilityScore)}`);
    console.info(`│  Best Practices Score    │ ${rpad(String(lh.bestPracticesScore), 7)}    │  ${scoreLabel(lh.bestPracticesScore)}`);
    console.info(`│  SEO Score               │ ${rpad(String(lh.seoScore), 7)}    │  ${scoreLabel(lh.seoScore)}`);
    console.info('├──────────────────────────┼────────────┼──────────────────┤');
    console.info(`│  LH FCP                  │ ${rpad(fmt(lh.firstContentfulPaint), 7)} ms │`);
    console.info(`│  LH LCP                  │ ${rpad(fmt(lh.largestContentfulPaint), 7)} ms │`);
    console.info(`│  LH Speed Index          │ ${rpad(fmt(lh.speedIndex), 7)} ms │`);
    console.info(`│  LH TTI                  │ ${rpad(fmt(lh.timeToInteractive), 7)} ms │`);
    console.info(`│  LH TBT                  │ ${rpad(fmt(lh.totalBlockingTime), 7)} ms │`);
    console.info(`│  LH CLS                  │ ${rpad(lh.cumulativeLayoutShift.toFixed(4), 7)}    │`);
    if (lh.opportunities.length > 0) {
      console.info('├──────────────────────────┴────────────┴──────────────────┤');
      console.info('│  Top Opportunities                                        │');
      lh.opportunities.slice(0, 5).forEach(op => {
        const savings = op.savingsMs ? ` (save ~${Math.round(op.savingsMs)} ms)` : '';
        console.info(`│    • ${op.title}${savings}`);
      });
    }
    console.info('├──────────────────────────────────────────────────────────┤');
    console.info(`│  LH Report : ${lh.htmlReportPath}`);
  }

  console.info('└──────────────────────────────────────────────────────────┘');
}

function scoreLabel(score: number): string {
  if (score >= 90) return 'GOOD';
  if (score >= 50) return 'NEEDS IMPROVEMENT';
  return 'POOR';
}

function fmt(ms: number): string {
  return ms > 0 ? Math.round(ms).toString() : 'N/A';
}

function rpad(s: string, width: number): string {
  return s.padEnd(width);
}

function safeName(name: string): string {
  return name.replace(/[^a-zA-Z0-9._-]/g, '_');
}

function tsStamp(): string {
  return new Date().toISOString().replace(/[:.]/g, '-').replace('T', '_').slice(0, 23);
}
